// cedar-os/dist/store/agentConnection/providers/custom.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/store/agentConnection/providers/custom.ts
var custom_exports = {};
__export(custom_exports, {
  customProvider: () => customProvider
});
module.exports = __toCommonJS(custom_exports);
var customProvider = {
  callLLM: async (params, config) => {
    if (config.config.callLLM) {
      return await config.config.callLLM(params, config);
    }
    throw new Error("Custom provider requires a callLLM function in config");
  },
  callLLMStructured: async (params, config) => {
    if (config.config.callLLMStructured) {
      return await config.config.callLLMStructured(params, config);
    }
    if (config.config.callLLM) {
      return await config.config.callLLM(params, config);
    }
    throw new Error(
      "Custom provider requires a callLLMStructured function or callLLM function in config"
    );
  },
  streamLLM: (params, config, handler) => {
    if (config.config.streamLLM) {
      return config.config.streamLLM(params, config, handler);
    }
    if (config.config.callLLM) {
      const abortController = new AbortController();
      const completion = (async () => {
        try {
          const response = await config.config.callLLM(params, config);
          handler({ type: "chunk", content: response.content || "" });
          handler({ type: "done", completedItems: [response.content || ""] });
        } catch (error) {
          handler({ type: "error", error });
        }
      })();
      return {
        abort: () => abortController.abort(),
        completion
      };
    }
    throw new Error(
      "Custom provider requires a streamLLM function or callLLM function in config"
    );
  },
  voiceLLM: async (params, config) => {
    if (config.config.voiceLLM) {
      return await config.config.voiceLLM(params, config);
    }
    return {
      content: ""
    };
  },
  handleResponse: async (response) => {
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    const data = await response.json();
    return {
      content: data.content || data.message || "",
      usage: data.usage,
      metadata: data.metadata || {}
    };
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  customProvider
});
