// cedar-os/dist/store/stylingSlice.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/store/stylingSlice.ts
var stylingSlice_exports = {};
__export(stylingSlice_exports, {
  createStylingSlice: () => createStylingSlice
});
module.exports = __toCommonJS(stylingSlice_exports);
var DEFAULT_STYLING = {
  darkMode: false,
  color: "#93C5FD",
  // blue-300
  secondaryColor: "#1D4ED8",
  // blue-700
  accentColor: "#FB923C"
  // orange-400
};
var createStylingSlice = (set) => ({
  // Initial state
  styling: DEFAULT_STYLING,
  // Actions
  setStyling: (newStyling) => set((state) => ({
    styling: { ...state.styling, ...newStyling }
  })),
  toggleDarkMode: () => set((state) => ({
    styling: { ...state.styling, darkMode: !state.styling.darkMode }
  }))
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createStylingSlice
});
