// cedar-os/dist/store/messages/renderers/createMessageRenderer.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/store/messages/renderers/createMessageRenderer.ts
var createMessageRenderer_exports = {};
__export(createMessageRenderer_exports, {
  createLegacyActionMessageRenderer: () => createLegacyActionMessageRenderer,
  createMessageRenderer: () => createMessageRenderer,
  createSetStateMessageRenderer: () => createSetStateMessageRenderer
});
module.exports = __toCommonJS(createMessageRenderer_exports);
function createMessageRenderer(p) {
  return p;
}
function createSetStateMessageRenderer(config) {
  const { namespace, setterKey, render, validateMessage } = config;
  const defaultValidate = (msg) => {
    if (msg.type !== "setState") return false;
    if (setterKey && msg.setterKey !== setterKey)
      return false;
    return true;
  };
  const rendererFn = (message) => {
    return render(message);
  };
  return {
    type: "setState",
    namespace,
    render: rendererFn,
    validateMessage: validateMessage ?? defaultValidate
  };
}
function createLegacyActionMessageRenderer(config) {
  const { namespace, setterKey, render, validateMessage } = config;
  const defaultValidate = (msg) => {
    if (msg.type !== "action") return false;
    if (setterKey && msg.setterKey !== setterKey)
      return false;
    return true;
  };
  const rendererFn = (message) => {
    return render(message);
  };
  return {
    type: "action",
    namespace,
    render: rendererFn,
    validateMessage: validateMessage ?? defaultValidate
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createLegacyActionMessageRenderer,
  createMessageRenderer,
  createSetStateMessageRenderer
});
